src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -1498,81 +1441,77 @@ std::vector<MergeTreeData::LoadPartResult> MergeTreeData::loadDataPartsFromDisk(
-                            thread_idx = *it;
+                            std::lock_guard lock{part_select_mutex};
+                            for (const auto & [_, node] : part->children)
+                                parts_to_load.push_back(node);
                         }
 
-                        auto & thread_parts = threads_parts[thread_idx];
-                        thread_part = thread_parts.back();
-                        thread_parts.pop_back();
-                        if (thread_parts.empty())
-                            remaining_thread_parts.erase(thread_idx);
-                    }
+                        {
+                            std::lock_guard lock(part_loading_mutex);
+                            loaded_parts.push_back(std::move(res));
+                        }
+                    }, 0
+                ));
+            }
 
-                    /// Pass a separate mutex to guard the set of parts, because this lambda
-                    /// is called concurrently but with already locked @data_parts_mutex.
-                    auto res = loadDataPartWithRetries(
-                        thread_part->info, thread_part->name, thread_part->disk,
-                        DataPartState::Active, part_loading_mutex, loading_parts_initial_backoff_ms,
-                        loading_parts_max_backoff_ms, loading_parts_max_tries);
+            /// Wait for all scheduled tasks
+            for (auto & future: parts_futures)
+                future.wait();
 
-                    thread_part->is_loaded = true;
-                    bool is_active_part = res.part->getState() == DataPartState::Active;
+            parts_futures.clear();
 
-                    /// If part is broken or duplicate or should be removed according to transaction
-                    /// and it has any covered parts then try to load them to replace this part.
-                    if (!is_active_part && !thread_part->children.empty())
-                    {
-                        std::lock_guard lock{part_select_mutex};
-                        for (const auto & [_, node] : thread_part->children)
-                            threads_parts[thread].push_back(node);
-                        remaining_thread_parts.insert(thread);
-                    }
-
-                    {
-                        std::lock_guard lock(part_loading_mutex);
-                        loaded_parts.push_back(std::move(res));
-                    }
-                }
-            });
+            /// It is now possible, that some other parts appeared in the queue for processing (parts_to_load)
+            /// So we need to recheck it.
         }
     }
     catch (...)
     {
-        /// If this is not done, then in case of an exception, tasks will be destroyed before the threads are completed, and it will be bad.
-        pool.wait();
+        /// Wait for all scheduled tasks
+        for (auto & future: parts_futures)
+            future.wait();
+
```

Comment:
Good catch. But I guess that the difference won't be so huge. This pool has N threads and a queue for N slots and producers will wait for workers in pool to complete tasks when the pool is full. In case of exception we would need to wait for only 2*N parts to be loaded. Is it Ok?

Reply:
Looks okay to me, it should work fine then.